### PR TITLE
Core/Support: Update the support/ticket system to 6.x

### PR DESCRIPTION
**Pull request by Carbenium**
Core/Mail: fix a logic error introduced in 829e11b6623f967672143d79cd8d31772f54866d

### CODE REVIEW

File: src/server/game/Support/SupportMgr.h [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+
+        return nullptr;
+    }
+
+    ComplaintTicketList GetComplaintsByPlayerGuid(ObjectGuid playerGuid) const
+    {
+        ComplaintTicketList ret;
+        for (auto const& c : _complaintTicketList)
+            if (c.second->GetPlayerGuid() == playerGuid)
+                ret.insert(c);
```
Comment (joschiwald) [inline src/server/game/Support/SupportMgr.h]: missing return

File: src/server/game/Support/SupportMgr.cpp [hunk 2]
```diff
@@ -0,0 +1,1031 @@
...
+        stmt->setUInt32(idx, _id);
+        stmt->setUInt32(++idx, lineIndex);
+        stmt->setUInt32(++idx, c.Timestamp);
+        stmt->setString(++idx, c.Text);
+
+        trans->Append(stmt);
+        ++lineIndex;
+    }
+    
+    CharacterDatabase.CommitTransaction(trans);
```
Comment (joschiwald) [inline src/server/game/Support/SupportMgr.cpp]: this won't work when enter SaveToDB with started transaction
```c++
bool isInTransaction = bool(trans);
if (!isInTransaction)
    trans = CharacterDatabase.BeginTransaction();

do stuff

if (!isInTransaction)
    CharacterDatabase.CommitTransaction(trans);
```

File: src/server/game/Handlers/TicketHandler.cpp [hunk 3]
```diff
@@ -42,140 +42,100 @@ void WorldSession::HandleGMTicketCreateOpcode(WorldPacket& recvData)
...
-        recvData >> count;
-
-        for (uint32 i = 0; i < count; i++)
+        if (packet.DataLength > 0 && packet.ChatHistoryData.DecompressedSize < 0xFFFF)
         {
-            uint32 time;
-            recvData >> time;
-            times.push_back(time);
-        }
+            ByteBuffer dest(packet.ChatHistoryData.DecompressedSize);
```
Comment (joschiwald) [inline src/server/game/Handlers/TicketHandler.cpp]: this will just do a reserve, you have to resize the bytebuffer

File: sql/updates/auth/2015_02_xx_xx_auth.sql [hunk 4]
```diff
@@ -0,0 +1,77 @@
+UPDATE `rbac_permissions` SET `name`='Command: reload support system' WHERE  `id`=650;
+
+DELETE FROM `rbac_permissions` WHERE `id` IN (799,800,801,802,803,804,805,806,807,808,809,810,811,812,813,814,815,816,817,818,819,820,821,822,823,824,825,826,827,828,829,830,831,832,833);
```
Comment (jackpoz) [inline sql/updates/auth/2015_02_xx_xx_auth.sql]: what was changed here to require to replace rbac_permissions ?

File: sql/updates/characters/2015_02_xx_xx_characters.sql [hunk 5]
```diff
@@ -0,0 +1,95 @@
+DROP TABLE IF EXISTS `lag_reports`;
+
+ALTER TABLE `gm_surveys`
+    RENAME TO `gm_survey`,
```
Comment (jackpoz) [inline sql/updates/characters/2015_02_xx_xx_characters.sql]: is there any need to rename the table ?

File: sql/updates/characters/2015_02_xx_xx_characters.sql [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    CHANGE COLUMN `overallComment` `comment` LONGTEXT NOT NULL AFTER `mainSurvey`;
+	
+ALTER TABLE `gm_subsurveys`
+    RENAME TO `gm_subsurvey`,
```
Comment (jackpoz) [inline sql/updates/characters/2015_02_xx_xx_characters.sql]: is there any need to rename the table ?

File: sql/updates/characters/2015_02_xx_xx_characters.sql [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    CHANGE COLUMN `subsurveyId` `questionId` INT(10) UNSIGNED NOT NULL DEFAULT '0' AFTER `surveyId`,
+	CHANGE COLUMN `rank` `answer` INT(10) UNSIGNED NOT NULL DEFAULT '0' AFTER `questionId`,
+	CHANGE COLUMN `comment` `answerComment` TEXT NOT NULL AFTER `answer`;
+
+
+ALTER TABLE `gm_tickets`
+    RENAME TO `gm_ticket`,
```
Comment (jackpoz) [inline sql/updates/characters/2015_02_xx_xx_characters.sql]: is there any need to rename the table ?

File: sql/updates/world/2015_02_xx_xx_world.sql [hunk 8]
```diff
@@ -0,0 +1,36 @@
+DELETE FROM `trinity_string` WHERE `entry` = 1190;
+INSERT INTO `trinity_string` (`entry`, `content_default`) VALUES (1190, 'Ticket system is disabled. You can\'t change its status.');
+
+UPDATE `command` SET `name`='reload support', `help`='Syntax: .reload support\r\nReload support system tables.' WHERE  `name`='reload gm_tickets';
+
+DELETE FROM `command` WHERE `name` IN ('ticket bug assign', 'ticket bug close', 'ticket bug closedlist', 'ticket bug comment', 'ticket bug delete', 'ticket bug list', 'ticket bug unassign', 'ticket bug view', 'ticket complaint assign', 'ticket complaint close', 'ticket complaint closedlist', 'ticket complaint comment', 'ticket complaint delete', 'ticket complaint list', 'ticket complaint unassign', 'ticket complaint view', 'ticket suggestion assign', 'ticket suggestion close', 'ticket suggestion closedlist', 'ticket suggestion comment', 'ticket suggestion delete', 'ticket suggestion list', 'ticket suggestion unassign', 'ticket suggestion view', 'ticket reset all', 'ticket reset gm', 'ticket reset bug', 'ticket reset complaint', 'ticket reset suggestion');
+INSERT INTO `command` (`name`, `permission`, `help`) VALUES
```
Comment (jackpoz) [inline sql/updates/world/2015_02_xx_xx_world.sql]: what was wrong with old data ?

File: src/server/game/Entities/Player/Player.cpp [hunk 9]
```diff
@@ -3244,17 +3244,12 @@ void Player::SendMailResult(uint32 mailId, MailResponseType mailAction, MailResp
     result.Command = mailAction;
     result.ErrorCode = mailError;
 
-    switch (mailError)
```
Comment (jackpoz) [inline src/server/game/Entities/Player/Player.cpp]: please submit each fix to its own PR, mails have nothing to do with tickets

File: src/server/game/Handlers/TicketHandler.cpp [hunk 10]
```diff
@@ -42,140 +42,101 @@ void WorldSession::HandleGMTicketCreateOpcode(WorldPacket& recvData)
...
-        {
-            uint32 pos = recvData.rpos();
             ByteBuffer dest;
-            dest.resize(decompressedSize);
+            dest.resize(size_t(packet.ChatHistoryData.DecompressedSize));
 
-            uLongf realSize = decompressedSize;
-            if (uncompress(dest.contents(), &realSize, recvData.contents() + pos, recvData.size() - pos) == Z_OK)
-            {
+            uLongf realSize = packet.ChatHistoryData.DecompressedSize;
```
Comment (jackpoz) [inline src/server/game/Handlers/TicketHandler.cpp]: not sure if we want to use uLongf type since we usually use only uint32/64/size_t

File: src/server/scripts/Commands/cs_go.cpp [hunk 11]
```diff
@@ -40,23 +40,26 @@ class go_commandscript : public CommandScript
...
-            { "grid",      rbac::RBAC_PERM_COMMAND_GO_GRID,      false, &HandleGoGridCommand,      "", NULL },
-            { "object",    rbac::RBAC_PERM_COMMAND_GO_OBJECT,    false, &HandleGoObjectCommand,    "", NULL },
-            { "taxinode",  rbac::RBAC_PERM_COMMAND_GO_TAXINODE,  false, &HandleGoTaxinodeCommand,  "", NULL },
-            { "trigger",   rbac::RBAC_PERM_COMMAND_GO_TRIGGER,   false, &HandleGoTriggerCommand,   "", NULL },
-            { "zonexy",    rbac::RBAC_PERM_COMMAND_GO_ZONEXY,    false, &HandleGoZoneXYCommand,    "", NULL },
-            { "xyz",       rbac::RBAC_PERM_COMMAND_GO_XYZ,       false, &HandleGoXYZCommand,       "", NULL },
-            { "ticket",    rbac::RBAC_PERM_COMMAND_GO_TICKET,    false, &HandleGoTicketCommand,    "", NULL },
-            { "",          rbac::RBAC_PERM_COMMAND_GO,           false, &HandleGoXYZCommand,       "", NULL },
-            { NULL,        0,                              false, NULL,                      "", NULL }
+            { "creature",           rbac::RBAC_PERM_COMMAND_GO_CREATURE,            false, &HandleGoCreatureCommand,                    "", NULL },
```
Comment (jackpoz) [inline src/server/scripts/Commands/cs_go.cpp]: any reason to modify this part of the sources ?

**Comment by jackpoz** [comment]
This PR has quite a lot of changes that look unneeded and/or unrelated, I'm also concerned about backward compatibility of the modified GM commands

**Comment by Aokromes** [comment]
Also,  Failed — The Travis CI build failed

**Comment by jackpoz** [comment]
Take a look at build warnings too while at it https://travis-ci.org/TrinityCore/TrinityCore/builds/51017962

**Comment by Carbenium** [comment]
Everything gets fixed since its WIP ;) Made the PR to have a travis build

File: sql/updates/characters/2015_02_xx_xx_characters.sql (hunk 5 above)
Comment (Carbenium) [inline sql/updates/characters/2015_02_xx_xx_characters.sql]: survey system will be done later. just adadpted all gm related tables to singular naming scheme.

File: sql/updates/characters/2015_02_xx_xx_characters.sql (hunk 7 above)
Comment (Carbenium) [inline sql/updates/characters/2015_02_xx_xx_characters.sql]: table names should be singular...

File: sql/updates/auth/2015_02_xx_xx_auth.sql (hunk 4 above)
Comment (Carbenium) [inline sql/updates/auth/2015_02_xx_xx_auth.sql]: new permissons were added. file has to be applyable multiple times, so delete is needed.

File: sql/updates/world/2015_02_xx_xx_world.sql (hunk 8 above)
Comment (Carbenium) [inline sql/updates/world/2015_02_xx_xx_world.sql]: again: new commands

File: src/server/scripts/Commands/cs_go.cpp (hunk 11 above)
Comment (Carbenium) [inline src/server/scripts/Commands/cs_go.cpp]: indentation would be wrong with new commands

**Comment by Carbenium** [comment]
@jackpoz: what do you mean with backward compatibility?

**Comment by jackpoz** [comment]
will all old gm commands still work with same syntax ?

**Comment by Carbenium** [comment]
everything except .reload support and .ticket reset... is compatible

**Comment by F03SD** [comment]
Use macos debug client for the try offsets.

**Comment by soulfrost** [comment]
where's problems of this PR ?

**Comment by jackpoz** [comment]
1) github says it cannot be merged automatically
2) no community feedback
3) https://github.com/TrinityCore/TrinityCore/pull/14169/files#r24798382

**Comment by Carbenium** [comment]
Will update it, as soon as i find some spare time

**Comment by Carbenium** [comment]
Updated to 6.1.0. Some community feedback would be great ;)

**Comment by et65** [comment]
I've tested all the system and everything is ok! Just a little problem, you prefixed all settings in "World.cpp" with "Support." and in "worldserver.conf.dist" with "Ticket.", just a little mistake to fix.

Good job :)!

**Comment by Carbenium** [comment]
Configs fixed. Thanks @et65

File: sql/base/characters_database.sql [hunk 12]
```diff
@@ -1619,93 +1619,218 @@ LOCK TABLES `gameobject_respawn` WRITE;
...
 /*!40101 SET character_set_client = utf8 */;
-CREATE TABLE `gm_subsurveys` (
-  `surveyId` int(10) unsigned NOT NULL AUTO_INCREMENT,
-  `subsurveyId` int(10) unsigned NOT NULL DEFAULT '0',
-  `rank` int(10) unsigned NOT NULL DEFAULT '0',
-  `comment` text NOT NULL,
-  PRIMARY KEY (`surveyId`,`subsurveyId`)
-) ENGINE=InnoDB DEFAULT CHARSET=utf8 COMMENT='Player System';
+CREATE TABLE `gm_bug` (
+	`id` INT(10) UNSIGNED NOT NULL,
```
Comment (Shauren) [inline sql/base/characters_database.sql]: thees tabs, tabssssssssss

**Comment by zwj-2014** [comment]
work fine. thank you man.
